src/Tes.Runner/Transfer/PipelineOptionsOptimizer.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Tes.Runner.Transfer
+{
+    /// <summary>
+    /// Pipeline options optimizer
+    /// </summary>
+    public class PipelineOptionsOptimizer
+    {
+        private readonly ISystemInfoProvider systemInfoProvider;
+        private const double MemoryBufferCapacityFactor = 0.4; // 40% of total memory
+        private const long MaxMemoryBufferSizeInBytes = BlobSizeUtils.GiB * 2; // 2 GiB of total memory
+        private const int MaxNumberOfWorkingOfThreads = 90;
```

Comment:
maybe rename to MaxWorkingThreadsCount.  I'm surprised it's not valuable to vary this based on core count?